lib/failure_detector.ml

```diff
@@ -0,0 +1,138 @@
+  in
+  Lwt_condition.wait cond
+
+(* TODO: An async thread should be started here, with a basic sleep, but not sure at all...
+     And we should return Either Ok Timeout *)
+let wait_ack_timeout t sequence_number timeout =
+  Lwt.pick
+    [
+      (let _ = Lwt_unix.sleep (Float.of_int timeout) in
+       Lwt.return @@ Result.Error "timeout");
+      (let _ = wait_ack t sequence_number in
+       Lwt.return @@ Result.ok "received ACK");
+    ]
+
+(* Basic Knuth shuffle => https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle *)
+let knuth_shuffle known_peers =
+  let initial_array = Array.length (Array.of_list known_peers) in
+  let shuffled_array = Array.copy (Array.of_list known_peers) in
+  for i = initial_array - 1 downto 1 do
+    let k = Random.int (i + 1) in
+    let x = shuffled_array.(k) in
+    shuffled_array.(k) <- shuffled_array.(i);
+    shuffled_array.(i) <- x
+  done;
+  Array.to_list shuffled_array
+
+(* Regarding the SWIM protocol, the list of peers is not ordered.
+   Hence, I basically went for a shuffle after adding the new peer *)
+let add_peer peer_to_add t =
+  let new_peers = peer_to_add :: t.peers in
+  let shuffled_list = knuth_shuffle new_peers in
+  t.peers <- shuffled_list
+
+(* Regarding the SWIM protocol, if peer A cannot get ACK from peer B (timeout):
+   A sets B as `suspicious`
+   A randomly picks one (or several, should it also be randomly determined?) peer(s) from its list
+   and ask him/them to ping B.*)
+(* This function return the random peer, to which we will ask to ping the first peer *)
+let rec pick_random_peers (peers : Peer.t list) number_of_peers =
+  match peers with
+  | [] -> failwith "pick_random_peers"
+  | elem :: list ->
+    if number_of_peers = 1 then
+      [elem]
+    else
+      elem :: pick_random_peers list (number_of_peers - 1)
+
+(* TODO: This function will update the status of the peer to Suspicious or Faulty *)
+let update_peer _t _client _peer = failwith "undefined"
+
+(* How should I use this function? Let the client of the library defines a `peer.init` using it at `msg_handler`? *)
+let[@warning "-32"] handle_payload t client (peer : Peer.t) msg =
+  let _ =
+    update_peer t client
+      Peer.{ status = Alive; socket_address = peer.socket_address } in
+  match msg with
+  | Ping -> send_acknowledge_to client peer
+  | PingRequest addr -> (
+    let seq_no = next_seq_no t in
+    let _ = send_ping_to client (Peer.retrieve_peer_from_address t.peers addr) in
+    match%lwt wait_ack_timeout t seq_no t.config.protocol_period with
+    (* TODO: I am not sure about using the backtick character, I think it makes it polymorphic, should we do something like that here? *)
+    | Ok _ -> Lwt.return ()
+    | Error _ -> send_acknowledge_to client peer)
+  | Acknowledge ->
+  match find t.acknowledges t.sequence_number with
+  | Some cond ->
+    Lwt_condition.broadcast cond ();
+    Lwt.return ()
+  | None ->
+    (* Unexpected ACK -- ignore *)
```

Comment:
If we get an ACK from a peer we didn't know before, can we actually just add it to the list of peers?

Reply:
This is fine, we don't want to crash just because of that 😅